### PR TITLE
refactor: locationSelectionViewController for better readability and modularity - WPB-9730

### PR DESCRIPTION
**Pull request by agisilaos**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-9730" title="WPB-9730" target="_blank">TaskWPB-9730</a>  refactor: locationSelectionViewController for better readability and modularity
  </td></table>
  <br />
 

This PR encompasses a series of refactorings to the LocationSelectionViewController aimed at enhancing code quality, readability, and maintainability. The primary focus has been on encapsulating functionalities related to map and location management within dedicated manager classes, thereby adhering to the Single Responsibility Principle and making the view controller more concise and focused.

### Introduction of MapViewController:

**Purpose**: To handle all map-related functionalities including the map view setup, annotation handling, and user location updates.

**Changes**:
- Created MapViewController class that encapsulates the MKMapView and its configurations.
- Moved map setup and interaction code from LocationSelectionViewController to MapManager.
- Implemented MapViewControllerDelegate within LocationSelectionViewController to handle map events like user location updates and region changes.

### Use of AppLocationManager:
**Purpose**: To manage all location services such as requesting location access, starting/stopping location updates, and handling location permissions.

**Changes**:
- Refactored LocationSelectionViewController to utilize AppLocationManager for all location services.
- Removed direct instances of CLLocationManager from LocationSelectionViewController and replaced them with references to AppLocationManager.shared.

### Refactoring Annotation Handling:
**Purpose**: To simplify annotation management by moving it to MapViewController.

**Changes**:
- Transferred the MKPointAnnotation and its view setup from LocationSelectionViewController to MapViewController.
- Added methods in MapViewController to manage adding and updating annotations on the map.

### Code Cleanup and Constraint Adjustments:
**Purpose**: To resolve crashes and issues arising from improper view hierarchy and constraints setup.

**Changes**:
- Adjusted auto layout constraints to avoid crashes related to non-existent common ancestors in the view hierarchy.
- Ensured that all UI components are added to the view and constraints are set up correctly before being activated.

### Improvements to Authorization Handling:
**Purpose**: To streamline how location permissions are handled.

**Changes**:
- Updated the didChangeAuthorization method to use AppLocationManager.shared to check and request permissions, reflecting a central management point for location services.

### Checklist

- [ ] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [ ] Description is filled and free of optional paragraphs.
- [ ] Adds/updates automated tests.

### CODE REVIEW

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/AppLocationManager.swift
Comment (johnxnguyen) [inline wire-ios/Wire-iOS/Sources/UserInterface/Location/AppLocationManager.swift]: How hard would it be to avoid using a singleton?

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/MapManager.swift
Comment (johnxnguyen) [inline wire-ios/Wire-iOS/Sources/UserInterface/Location/MapManager.swift]: It seems a bit strange that a manager class has a reference to a view. Perhaps instead of `MapManager` it could be a `MapViewController`?

Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/Location/MapManager.swift]: Done!

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/AppLocationManager.swift
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/Location/AppLocationManager.swift]: @johnxnguyen Done! https://github.com/wireapp/wire-ios/commit/1a8aaf2b62d71a5a770c1bb55833c1ea8f570bd4

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift
Comment (caldrian) [inline wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift]:
```suggestion
    private let mapViewController = MapViewController()
```

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/AppLocationManager.swift
Comment (markusfassbender) [inline wire-ios/Wire-iOS/Sources/UserInterface/Location/AppLocationManager.swift]:
```suggestion
final class AppLocationManager: NSObject {
```

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift
Comment (caldrian) [inline wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift]: Please pass the provided `animated` argument.

```suggestion
        if !userLocationAuthorized {
            mapViewController.mapView.restoreLocation(animated: animated)
        }
```

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/AppLocationManager.swift
Comment (markusfassbender) [inline wire-ios/Wire-iOS/Sources/UserInterface/Location/AppLocationManager.swift]: Hm, I don't see the benefit of passing `locationManager` in, as we don't mock it or use different configurations in code.
Can we just instantiate the object internally?

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift
Comment (caldrian) [inline wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift]: Please let's not use `UIScreen.safeArea`. It should be removed.
Can't we use something like `view.safeAreaInsets.bottom`?

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift
Comment (markusfassbender) [inline wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift]: Should this implementation `userLocationAuthorized` be moved to the manager?

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift
Comment (caldrian) [inline wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift:276]: I don't understand why the `LocationSelectionViewController` needs to update the `MapViewController`'s mapView. Isn't that exactly what the `MapViewController` is for?

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift
Comment (markusfassbender) [inline wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift]: The PR's description says:
> Created MapViewController class that encapsulates the MKMapView and its configurations.

So I am asking myself if more logic should be moved to the `MapViewController` to avoid explicit usage of types like `MKCoordinateRegion` outside.

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift
Comment (markusfassbender) [inline wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift]: The PR's description says:
> Created MapViewController class that encapsulates the MKMapView and its configurations.

So I am asking myself if it's enough to provide an encapsulate `UIView` and avoid to expose the type `MKMapView`, because this can be configured in any way and can skip the controller. One example from line 326:
`mapViewController.mapView.showsUserLocation = true`

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift
Comment (caldrian) [inline wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift]: Why is setup code for the mapView in the `LocationSelectionViewController` and not in the `MapViewController`?

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/MapViewController.swift
Comment (caldrian) [inline wire-ios/Wire-iOS/Sources/UserInterface/Location/MapViewController.swift]:
```suggestion
final class MapViewController {
```

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/MapViewController.swift
Comment (caldrian) [inline wire-ios/Wire-iOS/Sources/UserInterface/Location/MapViewController.swift:40]: I think it's weird that we have this view controller, which is actually never presented and doesn't use its `view`, and also that it owns another view (`mapView`) which is presented somewhere else.

**Review by caldrian — COMMENTED** [review]
Nice documentation in the PR description.
Could we use at least parts of those texts as doc comments on the classes?

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/MapViewController.swift
Comment (caldrian) [reply wire-ios/Wire-iOS/Sources/UserInterface/Location/MapViewController.swift]: Actually the name is confusing, since it's not a subclass of `UIViewController`.

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift]: @caldrian For some reason using `view.safeAreaInsets.bottom` completely broke the UI that's why I reverted the change back to `UIScreen.safeArea`

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift
Comment (caldrian) [reply wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift:276]: Sorry, I got confused by the name `MapViewController`. I thought it's a subclass of `UIViewController`.

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/MapViewController.swift
Comment (caldrian) [reply wire-ios/Wire-iOS/Sources/UserInterface/Location/MapViewController.swift:40]: Sorry, I got confused by the name MapViewController. I thought it's a subclass of UIViewController.

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/AppLocationManager.swift
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/Location/AppLocationManager.swift]: Fixed: https://github.com/wireapp/wire-ios/commit/a0f70c0a4d69130334fcb3b07317f00d0bfd66a9

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/MapViewController.swift
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/Location/MapViewController.swift]: @caldrian Do you have any better name alternatives?

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift]: Done: https://github.com/wireapp/wire-ios/commit/8d186be1f4bab13dfe60eff797bd7693ec9cfa30

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/MapViewController.swift
Comment (johnxnguyen) [reply wire-ios/Wire-iOS/Sources/UserInterface/Location/MapViewController.swift]: Shouldn't it be a subclass of `UIViewController` though?

Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/Location/MapViewController.swift]: Done: https://github.com/wireapp/wire-ios/commit/af7d762cc7c5ef90d810bdeee827c5d0749fa2de

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift]: Fixed it!

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/AppLocationManager.swift
Comment (johnxnguyen) [inline wire-ios/Wire-iOS/Sources/UserInterface/Location/AppLocationManager.swift]: I'm wondering what the purpose of `AppLocationManager` is as it appears to wrap `CLLocationManager` without adding anything extra.

Comment (caldrian) [reply wire-ios/Wire-iOS/Sources/UserInterface/Location/AppLocationManager.swift]: I actually like it, since it allows for mocking the `CLLocationManager`.
I would have chosen a different name, though.
With `UserNotificationCenterAbstraction` and `UserNotificationCenterWrapper` I did the same.

Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/Location/AppLocationManager.swift]: That was my intention. A better separation of concerns @johnxnguyen

Comment (johnxnguyen) [reply wire-ios/Wire-iOS/Sources/UserInterface/Location/AppLocationManager.swift]: But doesn't it need to be abstracted with a protocol to allow mocking? Or maybe I'm missing something.

Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/Location/AppLocationManager.swift:31]: As a protocol I introduced `AppLocationManagerDelegate`. Did you have anything else in mind?

```
protocol AppLocationManagerDelegate: AnyObject {

    func didUpdateLocations(_ locations: [CLLocation])

    func didFailWithError(_ error: Error)

    func didChangeAuthorization(status: CLAuthorizationStatus)

}
```

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift]: Done: https://github.com/wireapp/wire-ios/commit/4d243f4f3027aa40b2855efc4b5ac38e95552965

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/AppLocationManager.swift
Comment (caldrian) [reply wire-ios/Wire-iOS/Sources/UserInterface/Location/AppLocationManager.swift]: I agree, a protocol is missing, I haven't noticed it.

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift
Comment (markusfassbender) [reply wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift]: `view.safeAreaInsets.bottom` usually only contains valid Constants after the view has been added to the screen (`viewDidAppear`). This sometimes causes issues with `constant`.

Better would be to use `safeAreaLayoutGuide.bottomAnchor`.

**Review by markusfassbender — APPROVED** [review]

**Review by caldrian — APPROVED** [review]
approving now, but it would be nice to see a protocol for the AppLocationManager before merging
great work!

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/AppLocationManager.swift
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/Location/AppLocationManager.swift]: Done: https://github.com/wireapp/wire-ios/pull/1377/commits/681fc1f6c82b786cc8ecf862a7c90e67c3c01c03

@johnxnguyen @caldrian

File: wire-ios/Wire-iOS/Resources/Base.lproj/Localizable.strings
Comment (johnxnguyen) [inline wire-ios/Wire-iOS/Resources/Base.lproj/Localizable.strings:400]: Are these new strings? Did they come from product?

Comment (agisilaos) [reply wire-ios/Wire-iOS/Resources/Base.lproj/Localizable.strings:400]: Yep! Astrid provided them!

**Review by johnxnguyen — APPROVED** [review]

File: wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/Location/LocationSelectionViewController.swift]: Done!

**Pull request merged**
